File: Data Structures & Algorithms/combination-target-sum/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        let res = []
        this.backtrack(nums, 0, target, [], res)
        return res
    }

    backtrack(nums, i, remain, subset, res) {
        if (remain === 0) {
            res.push([...subset])
            return
        } else if (i >= nums.length || remain < 0) {
            return
        }

        subset.push(nums[i])
        this.backtrack(nums, i, remain - nums[i], subset, res)
        subset.pop()
        this.backtrack(nums, i+1, remain, subset, res)
    }
}

File: Data Structures & Algorithms/max-water-container/submission-2.go
func maxArea(nums []int) int {
    res := 0
    l,r  := 0, len(nums)-1
    
    for l < r {
        if nums[l] > nums[r] {
            if (nums[r] * (r-l)) > res {
                res = nums[r] * (r-l)
            }
            r--
        } else  {
            if (nums[l] * (r - l)) > res {
                res = nums[l] * (r - l)
            }
            l++
        }
    }

    return res
}
